Feature/trace-collector-service

sankalpatimilsina12 opened the pull request:
This adds a dockerized trace collection service.

1. The directory has been renamed from the original `ndntdump-exp-2023` to generic `traces` which will be present on the host and will contain collected traces.
2. The image is under GCR, tagged and is under a username, happy to upload elsewhere.

Pesa commented on docker-compose.yml:
```diff
@@ -222,12 +222,12 @@ services:
 
   trace-collector:
-    image: sankalpatimilsina/scheduler:latest
+    image: ghcr.io/sankalpatimilsina12/trace-collector:20250214
```
I can't find this on your account, where is it? (and its associated source repo)

sankalpatimilsina12 replied:
It seems like it was private after I uploaded it. The image is [public ](https://github.com/users/sankalpatimilsina12/packages/container/package/trace-collector) now, but I just discovered that the namespace (user / organization) for the source repository and container image on Github must be same for them to be linked. I will need to reupload/retag it.

Pesa replied:
ok I can see it now. I wouldn't worry too much about the name for now, as long as the reviewers can see the code publicly somewhere. Eventually we may want to move the repo under https://github.com/tntech-ngin with the rest, or even directly in the testbed repo (up to the maintainers).

sankalpatimilsina12 replied:
Sounds good, I added the source repository to the original PR description.

Pesa reviewed (commented):
Please squash the commits.

yoursunny commented on docker-compose.yml:
```diff
@@ -220,3 +220,17 @@ services:
+    volumes:
+      - /home/.ssh:/root/.ssh:ro
+      - /home/ndnops/traces:/dump
+      - ${PWD}/dist/nlsr:/config
```
Do you really need write access to the `${PWD}/dist/nlsr` folder?

sankalpatimilsina12 replied:
I was trying to figure this out, we need the `nlsr.conf` file.

Pesa replied:
This was a semi-hack we had in the old scripts to get a unique "node name", IIRC. I think we can do better now with the new framework, e.g. pass the name to the container via env or something...

Pesa replied:
@pulsejet what's the easiest way to pass the `site` from host_vars to the container?

pulsejet replied:
Create a new config template, have the container read off it.

sankalpatimilsina12 replied:
The latest commit has an `entrypoint.sh.j2` template which directly reads site name.

yoursunny commented on docker-compose.yml:
```diff
@@ -220,3 +220,17 @@ services:
+    init: true
+    network_mode: service:nfd
+    volumes:
+      - /home/.ssh:/root/.ssh:ro
```
`/home/.ssh` seems wrong.

yoursunny replied:
The Docker image contains neither `ssh` nor `scp` command. I'm uncertain how you are going to use the SSH keys.

sankalpatimilsina12 replied:
Yes, the path is for sure wrong, now that I see it 😄

sankalpatimilsina12 replied:
The latest image has the relevant commands included.

yoursunny commented:
As discussed in 2025-02-28 meeting, the SSH key pair should be handled as follows:

* trace-collector should not use the SSH key pair that belongs to the operator or assume its existence.
* trace-collector should mount a Docker volume that contains its own SSH key pair (for convenience, it could be mounted at `/root/.ssh`).
* The public key of the trace-collector's key pair should be published through file-server. This could be achieved by mounting a sub-directory of what's served by the file-server (for example, `${PWD}/dist/file-server/trace-collector-pubkey`).

Upon service start, the entrypoint script should perform the following:

1. If the SSH key pair does not exist in the mountpoint of the Docker volume, generate a key pair noninteractively.
2. Copy the public key to a directory published by the file-server. This should be done unconditionally, even if the key pair wasn't generated in this session, because it is possible that last time the container crashed between key pair generation and copying.
3. Invoke the main script.

After the first run of the service, the public key of a router shall be retrievable with [ndncat command](https://github.com/yoursunny/NDNts/tree/61f8da34bb3a5217abd78ae5b2f7ad28a0708522/pkg/cat):

```bash
ndncat get-segmented /ndn/edu/neu/file-server/trace-collector-pubkey/id_ed25519.pub
```

sankalpatimilsina12 commented:
The `entrypoint.sh.j2` has the included flow as discussed. I have added two persistent volume claims for dump and ssh storage, which were added for preserving keys and dumps. The flow was locally verified at my end.

1. Site is read in the template and cleaned. Eg. `/edu/memphis` get converted to `edu_memphis`.
2. Ssh keys are generated if not already present and copied to fileserver's subdirectory for trace-collector.
3. Main script is then invoked with the cleaned site name.

Pesa commented on docker-compose.yml:
```diff
@@ -220,3 +220,26 @@ services:
+    restart: unless-stopped
+    depends_on:
+      nfd: { condition: service_healthy }
+      file-server: { condition: service_healthy }
```
is there an actual dependency on the file server?

sankalpatimilsina12 replied:
I put this there I think from a documentary perspective, there is no actual dependency. It should be just `nfd`.

sankalpatimilsina12 commented on templates/trace-collector/entrypoint.sh.j2:
```diff
@@ -0,0 +1,31 @@
+SSH_DIR="/root/.ssh"
+PRIVATE_KEY="$SSH_DIR/id_rsa"
+PUBLIC_KEY="$SSH_DIR/id_rsa.pub"
+PUBKEY_EXPORT_PATH="/file-server/trace-collector-pubkey/id_rsa.pub"
```
The file name has been referenced multiple times in the script: Checking existence, non-interactive generation with the file name specified, and during copying. Should we keep the private and public key variables? The export path I believe can just be a directory name.

Pesa replied:
> Should we keep the private and public key variables? 

If you prefer to use a variable, that's fine. But one variable for the private key should be enough, the public key name should always be privatekey + `.pub` per openssh conventions.

> The export path I believe can just be a directory name.

On second thought, we should really avoid using a dot in the NDN name of the pubkey (e.g. `id_rsa.pub`) because one dot becomes four dots in NDN URI syntax, which is quite ugly. I'd say just use the directory name `trace-collector-pubkey` as the actual key name. I don't think we need a subdir for this.

yoursunny replied:
> one dot becomes four dots in NDN URI syntax

False.
One dot becomes four dots only if the component consists solely of dots.

sankalpatimilsina12 replied:
> If you prefer to use a variable, that's fine. But one variable for the private key should be enough, the public key name should always be privatekey + `.pub` per openssh conventions.

Sure.
